### DISCUSSION TITLE
Dynamic line graph in real time

### DISCUSSION THREAD

**Discussion by Jalvearc** [h2oai/wave#1769] [Q&A] [answered]
Hello, I request help because I tried everything to achieve this problem and it was not possible.

My problem is I need to develop an hour by hour graph that shows me the results.
the examples that are in the documentation show static values ​​and not any that interest me to guide me
![image](https://user-images.githubusercontent.com/120686730/207975808-57d76fe1-ef80-438e-9b1f-3f12d1b0524a.png)
This is my data frame, the only value that is going to change is the one that is last, once the hour passes, it will start updating at the next hour.
Can someone help me or give an example to create this interactive graph, I would really appreciate it, I am really stuck with this problem.

**Comment by Jalvearc** [comment]
Hello I have advanced this but I can not make it work can someone guide me please

![image](https://user-images.githubusercontent.com/120686730/208008071-68f94971-07d0-424d-af02-0aecd080a33d.png)

**Comment by mturoci** [comment, marked as answer]
Hey @Jalvearc! Try having a look at [data buffers](https://wave.h2o.ai/docs/buffers/).

**Reply by Jalvearc** [reply]
Hello, thanks for answering, I have reviewed and reviewed and no matter how hard I try, I can't get interested in the last value of the data frame, that is, I can have 10,2,4,5,6 and in the following minutes I will have 10,2, 4,5,9 but I can't get the graph to take the buffer, I don't know if I'm indexing the list or the dictionary wrong

**Reply by mturoci** [reply]
Can you share your code that I could run myself? With static mocked data ideally to keep things simple.

**Reply by Jalvearc** [reply]
page = site['/monitor']


h =page.add('example8', ui.plot_card(
    box='1 8 8 5',
    title='Line',
    data=data(fields='day value', rows=dict(usage=0.0)),
    plot=ui.plot([ui.mark(type='line', x='=day', y='=value', y_min=0)])
))



tick = 0
while True:
    tick += 1
    
    df=pd.DataFrame()
    df['day']=['20-12','20-13','20-14','20-15']
    df['value']=[10,20,10,5]
      
    plotdata= df.values.tolist()
    h.data.usage[-1]=[plotdata]
    
    page.save()
    time.sleep(1)

**Reply by Jalvearc** [reply]
I am attaching the piece of code that I need, for example, those values ​​would come from a database that was updated, so they would change, but the idea is that with those values, graph the line graph


my idea is this

I got the values

20-12:10
20-13:20
20-14:10
20-15 : 5

I graph the line graph in this second but thirty seconds later this will happen
20-12:10
20-13:20
20-14:10
20-15 : 25

I need it to detect the change or later this can happen

20-12:10
20-13:20
20-14:10
20-15 : 25
20-16:5
I understood the buffer but implementing it was difficult for me

thanks for your help

**Reply by mturoci** [reply]
Unfortunately, the code you shared is not runnable + it could use some formatting to make our lives easier. Btw, did you manage to render the initial line plot correctly? Without the subsequent updates.

**Reply by Jalvearc** [reply]
page = site['/ayuda']

df=pd.DataFrame()
df['day']=['20-12','20-13','20-14','20-15','20-16']
df['value']=[10,20,10,5,17]


c=data(fields='day value', rows=df.values.tolist())

a=page.add('example', ui.plot_card(
    box='1 1 4 5',
    title='Line',
    data=c,
    plot=ui.plot([ui.mark(type='line', x='=day', y='=value', y_min=0)])
))

   
page.save()

**Reply by Jalvearc** [reply]
This is the graph with the static values ​​represented by the line graph

![image](https://user-images.githubusercontent.com/120686730/208114287-38073e31-f827-462c-b768-f4f0354e4e1c.png)

**Reply by Jalvearc** [reply]
Now what I have not been able to do is leave that in the while so that each modification is represented in the line graph

**Reply by mturoci** [reply]
This looks better. Seems like the most suitable buffer for your case would be `cyclic` (FIFO) one. For that to work, you need to specify `size` attr in your `data` function and then simply append the new data once it arrives.

```py
# Plot / Line
# Make a line #plot.
# ---
import time
from h2o_wave import site, data, ui
import pandas as pd

page = site['/ayuda']

df = pd.DataFrame()
df['day'] = ['20-12', '20-13', '20-14', '20-15', '20-16']
df['value'] = [10, 20, 10, 5, 17]

page['example'] = ui.plot_card(
    box='1 1 4 5',
    title='Line',
    # Use cyclic buffer with size=-5 to keep only last 5 values. Change as needed.
    data=data(fields='day value', size=-5, rows=df.values.tolist()),
    plot=ui.plot([ui.mark(type='line', x='=day', y='=value', y_min=0)])
)

# Render the initial plot.
page.save()

# Update new data (can be in while loop)
time.sleep(5)
page['example'].data[-1] = ['20-17', 20]  # append new data
page.save()

```

**Reply by Jalvearc** [reply]
This was wonderful, I just had to make a few small adjustments to the code so that it would work for me in real time and testing the information, I appreciate your help, it was very useful, I already had many hours dedicated to this problem and I can continue advancing, thank you

**Reply by mturoci** [reply]
Glad I could help :)
